### PR TITLE
[Hybrid Parallel] Add model parallel support in dygraph

### PR DESCRIPTION
**Pull request by ForFishes**
### PR types

New features
### PR changes

APIs
### Describe

[Hybrid Parallel] Add model parallel support in dygraph,

### CODE REVIEW

File: paddle/fluid/framework/distributed_strategy.proto [hunk 1]
```diff
@@ -43,6 +43,12 @@ message ShardingConfig {
   optional int32 pp_degree = 11 [ default = 1 ];
 }
 
+message HybridConfig {
+  optional int32 num_data_parallel = 1 [ default = -1 ];
+  optional int32 num_model_parallel = 2 [ default = 1 ];
+  optional int32 num_pipeline_parallel = 3 [ default = 1 ];
```
Comment (sandyhouse) [inline paddle/fluid/framework/distributed_strategy.proto]: Use dp_degree, mp_degree and pp_degree to be consistent with static graph.

File: python/paddle/distributed/fleet/__init__.py [hunk 2]
```diff
@@ -21,6 +21,7 @@
 from .data_generator import MultiSlotDataGenerator, MultiSlotStringDataGenerator
 from . import metrics
 from .base.topology import CommunicateTopology, HybridCommunicateGroup
+from .mpu import random, layers
```
Comment (sandyhouse) [inline python/paddle/distributed/fleet/__init__.py]: why import these two files?

File: python/paddle/distributed/fleet/base/distributed_strategy.py [hunk 3]
```diff
@@ -867,6 +867,62 @@ def pipeline_configs(self, configs):
...
+
+        Examples:
+          .. code-block:: python
+            import paddle.distributed.fleet as fleet
+            strategy = fleet.DistributedStrategy()
+            strategy.hybrid_configs = {
+                "num_data_parallel": 1,
+                "num_model_parallel": 2,
+                "num_pipeline_parallel": 1}
+        """
```
Comment (sandyhouse) [inline python/paddle/distributed/fleet/base/distributed_strategy.py:895]: use dp_degree, mp_degree and pp_degree

File: python/paddle/distributed/fleet/base/distributed_strategy.py [hunk 4]
```diff
@@ -867,6 +867,62 @@ def pipeline_configs(self, configs):
...
+
+        return get_msg_dict(self.strategy.model_parallel_configs)
+
+    @model_parallel_configs.setter
+    def model_parallel_configs(self, configs):
+        check_configs_key(self.strategy.model_parallel_configs, configs,
+                          "model_parallel_configs")
+        assign_configs_value(self.strategy.model_parallel_configs, configs)
+
```
Comment (sandyhouse) [inline python/paddle/distributed/fleet/base/distributed_strategy.py]: delete unnecessary model_parallel_configs in this pr.

File: python/paddle/distributed/fleet/base/fleet_base.py [hunk 5]
```diff
@@ -234,6 +235,40 @@ def init(self, role_maker=None, is_collective=False, strategy=None):
...
+        self.mp_num = self.hybrid_configs["num_model_parallel"]
+        self.pp_num = self.hybrid_configs["num_pipeline_parallel"]
+
+        if self.dp_num < 0:
+            nranks = paddle.distributed.get_world_size()
+            self.dp_num = nranks // (self.mp_num * self.pp_num)
+
+        self._topology = tp.CommunicateTopology(
+            hybrid_group_names=["data", "model", "pipe"],
+            dims=[self.dp_num, self.mp_num, self.pp_num])
```
Comment (sandyhouse) [inline python/paddle/distributed/fleet/base/fleet_base.py]: use ['data', 'pipe', 'model'] instead.

File: python/paddle/distributed/fleet/mpu/layers.py [hunk 6]
```diff
@@ -0,0 +1,193 @@
...
+
+        per_part_size = (
+            num_embeddings + self.world_size - 1) // self.world_size
+        last_part_size = num_embeddings - per_part_size * (self.world_size - 1)
+        if self.rank == self.world_size - 1:
+            per_part_size = last_part_size
+        per_part_size += 1  # make the last row as the padding index
+        self.per_part_size = per_part_size
+
+        with get_rng_state_tracker().rng_state():
```
Comment (sandyhouse) [inline python/paddle/distributed/fleet/mpu/layers.py]: delete unused with

File: python/paddle/distributed/fleet/mpu/layers.py [hunk 7]
```diff
@@ -0,0 +1,193 @@
...
+        assert out_features % self.world_size == 0, (
+            "Number of column of the weight for linear ({}) must be"
+            " divisible by model parallel size ({})".format(out_features,
+                                                            self.world_size))
+        self.output_size_per_partition = out_features // self.world_size
+
+        self._weight_attr = weight_attr
+        self._dtype = self._helper.get_default_dtype()
+
+        with get_rng_state_tracker().rng_state():
```
Comment (sandyhouse) [inline python/paddle/distributed/fleet/mpu/layers.py]: same as above

File: python/paddle/distributed/fleet/mpu/layers.py [hunk 8]
```diff
@@ -0,0 +1,193 @@
...
+        self.rank = tp._HYBRID_PARALLEL_GROUP.get_model_parallel_rank()
+
+        assert in_features % self.world_size == 0, (
+            "Number of row of the weight for linear ({}) must be"
+            " divisible by model parallel size ({})".format(in_features,
+                                                            self.world_size))
+
+        self.input_size_per_partition = in_features // self.world_size
+
+        with get_rng_state_tracker().rng_state():
```
Comment (sandyhouse) [inline python/paddle/distributed/fleet/mpu/layers.py]: same as above

**Review by sandyhouse — CHANGES_REQUESTED** [review]

File: paddle/fluid/framework/distributed_strategy.proto (hunk 1 above)
Comment (ForFishes) [reply paddle/fluid/framework/distributed_strategy.proto]: done

File: python/paddle/distributed/fleet/__init__.py (hunk 2 above)
Comment (ForFishes) [reply python/paddle/distributed/fleet/__init__.py]: rm it

File: python/paddle/distributed/fleet/base/distributed_strategy.py (hunk 3 above)
Comment (ForFishes) [reply python/paddle/distributed/fleet/base/distributed_strategy.py:895]: done

File: python/paddle/distributed/fleet/base/distributed_strategy.py (hunk 4 above)
Comment (ForFishes) [reply python/paddle/distributed/fleet/base/distributed_strategy.py]: ok, fix it

File: python/paddle/distributed/fleet/base/fleet_base.py (hunk 5 above)
Comment (ForFishes) [reply python/paddle/distributed/fleet/base/fleet_base.py]: done

File: python/paddle/distributed/fleet/mpu/layers.py (hunk 6 above)
Comment (ForFishes) [reply python/paddle/distributed/fleet/mpu/layers.py]: rm it

File: python/paddle/distributed/fleet/mpu/layers.py (hunk 7 above)
Comment (ForFishes) [reply python/paddle/distributed/fleet/mpu/layers.py]: fix it

File: python/paddle/distributed/fleet/mpu/layers.py (hunk 8 above)
Comment (ForFishes) [reply python/paddle/distributed/fleet/mpu/layers.py]: done

**Review by XieYunshen — DISMISSED** [review]

**Review by sandyhouse — DISMISSED** [review]
LGTM

**Review by sandyhouse — APPROVED** [review]
LGTM

**Review by XieYunshen — APPROVED** [review]

**Review by fuyinno4 — APPROVED** [review]

**Pull request merged**